src/arch/MovieTexture/MovieTexture_FFMpeg.cpp

```diff
@@ -265,114 +302,127 @@ int MovieDecoder_FFMpeg::SendPacketToBuffer()
 
 	while (true)
 	{
-		int ret = avcodec::av_read_frame(av_format_context_, frame_buffer_.back()->packet);
+		int ret = avcodec::av_read_frame(av_format_context_, packet_buffer_.back()->packet);
 		/* XXX: why is avformat returning AVERROR_NOMEM on EOF? */
 		if (ret < 0)
 		{
-			/* EOF. */
 			end_of_file_ = 1;
 			return 0;
 		}
 
-		if (frame_buffer_.back()->packet->stream_index == av_stream_->index)
+		if (packet_buffer_.back()->packet->stream_index == av_stream_->index)
 		{
-			current_packet_offset_ = 0;
 			return 1;
 		}
 		/* It's not for the video stream; ignore it. */
-		avcodec::av_packet_unref(frame_buffer_.back()->packet);
+		avcodec::av_packet_unref(packet_buffer_.back()->packet);
 	}
 }
 
-int MovieDecoder_FFMpeg::DecodePacketInBuffer() {
+int MovieDecoder_FFMpeg::DecodePacketToFrame() {
 	if (cancel_) {
 		return -2;
 	}
-	if (end_of_file_ == 0 && current_packet_offset_ == -1) {
-		return 0; /* no packet */
+
+	frame_buffer_position_ %= frame_buffer_.size();
+	packet_buffer_position_ %= total_frames_;
+	FrameHolder* frame = frame_buffer_[frame_buffer_position_].get();
```

Comment:
```
FrameHolder* frame = std::next(frame_buffer_.begin(), frame_buffer_position_)->get();
PacketHolder* packet = std::next(packet_buffer_.begin(), packet_buffer_position_)->get();
```

Reply:
> ```
> FrameHolder* frame = std::next(frame_buffer_.begin(), frame_buffer_position_)->get();
> PacketHolder* packet = std::next(packet_buffer_.begin(), packet_buffer_position_)->get();
> ```

For me it seems like you want each time reset iterator to begin, traverse whole list until the frame_buffer_position and only then get next. For list such operation is heavy.

If a buffer is a known fixed size, having just vector in memory and loop it via `iteration % size` looks faster and better. Or list should be linked last -> first, so we can endlessly iterate in it.

Reply:
That's true, but also it's entirely possible `vector` is superior to `list` for this operation, we'd have to try both to know for sure.

Reply:
> If a buffer is a known fixed size, having just vector in memory and loop it via `iteration % size` looks faster and better. Or list should be linked last -> first, so we can endlessly iterate in it.

The looping linked list is what I had in mind. Most of the trouble I had was with looping the movie in place/resetting correctly, so I figure a list would help mostly in eliminating potential off by one errors and keeping stronger continuity.

That being the case, it's not impossible to do via vectors and modulo'ing in the right spot, which is why I figure the code is potentially "good enough" for now.